Config API

jagerman opened the pull request.

jagerman committed:
Add writeup of config merge logic

majestrate commented on docs/config-merge-logic.md:
```diff
@@ -0,0 +1,576 @@
+
+Clients ignore updates according to two criteria:
+
+- "within five": a client ignores any update with a seqno value that is not within the last five
```
note why it's 5. 

aka: this value controls conflict resolution depth via [insert metric here] so we chose 5 because ass number.

jagerman replied:
This is chosen by the caller now.

majestrate commented on docs/config-merge-logic.md:
```diff
@@ -0,0 +1,576 @@
+  smaller one is contained within the diff of the larger one then the smaller one is stale and will
+  be ignored.
+
+## Simple updates
```
stylistic nitpick: reword heading to be `non conflicting updates`

jagerman committed:
(squash me) updates:

- clarify that deletions don't affect anything
- add encryption, using deterministic nonce
- change the "within 5" rule to "before N" so that different message
  types can choose how much conflict they want to allow (and thus how
  much diff overhead they want to accept).

ghost commented:
Is semantic patch a useful concept for the config merge discussion?

There are some semantic patch implementations for different languages, like [Coccinelle for C ](https://coccinelle.gitlabpages.inria.fr/website/), or [JSON Patch](https://jsonpatch.com/). I see some overlaps here, hopefully, we can borrow some useful ideas or reuse some useful tools.

ghost commented:
Is there any plan to prevent a buggy client from propagating a malformed configuration to other clients? I'm guessing a local client might want to do a dry run and [validate the schema](https://json-schema.org/) before actually applying the update from a remote client?

majestrate commented:
we aren't using json for serialization as it is too permissive, not compact enough and the syntax is excessively complicated. we are using bittorrent encoding, see [BEP0003](https://www.bittorrent.org/beps/bep_0003.html#bencoding) bencode section.

majestrate commented:
additionally json patch doesnt provide enough conflict resolution for what we are looking to do. it does a single depth diff but we need a LOT more than that.

jagerman commented:
JSON isn't suitable for encoding in the first place because:
- binary data gets bloated by having to go through hex (2x size) or base64 (1.33x size) or base85 (1.25x size), and we expect to have a fair amount of binary data (keys, nonces, encrypted values, etc.) stored in config messages.
- given data doesn't have a unique encoding (e.g. `{"foo": 123}` and `{"foo":123}` are the same data, but different encodings; strings also have many possible equivalent encoded forms, e.g. `"foo\.bar"`, `"foo.bar"`, `"\u0066oo.bar"`, etc.).  Bencoding has exactly one valid representation of any given data structure, which is crucial here (because we rely on storage-server de-duping of identical stored data to not have to worry about clients racing to push the same content).

> I'm guessing a local client might want to do a dry run and [validate the schema](https://json-schema.org/) before actually applying the update from a remote client?

Yes: clients will ignore any incoming config message that doesn't conform to the requirements.  But we really want to be confident in the base layer that no such buggy content gets pushed (which is why all the Session clients will be using this library to produce messages) because if there was such a buggy version producing invalid data, that client would likely also be merging and propagating as non-buggy clients push data.  (This wouldn't be catastrophic in that it creates a runaway publishing loop, but would result in a perpetual split of the data that two different client versions would see).

fracting commented on docs/config-merge-logic.md:
```diff
@@ -0,0 +1,609 @@
+            "good": [99, 123, "Foo", "bar"],
+            "int1": 42,
+            "int2": 2,
+            "string2": "hello"
```
If I understand correctly, we are missing `"string3": "omg"` here?

jagerman replied:
I spotted this and some other errors in here when using these as test cases, which are updated.

fracting commented on docs/config-merge-logic.md:
```diff
@@ -0,0 +1,609 @@
+   - the "current" diff of every message is added to the replay sets (the seqno and hash are
+     extracted and computed from the message itself).
+   - past diffs in the messages (i.e. from the `"<"` list) are added to the replay set, *if* they
+     are not yet in the reply set, and have a seqno >= newseqno-5.  This should be done in order
```
typo: `reply set` should be `replay set`

jagerman replied:
Thanks, fixed.

majestrate commented on docs/config-merge-logic.md:
```diff
@@ -0,0 +1,609 @@
+- strings, which are of max length 4096 bytes.
+- sets of one or more values, which may contain unique integer or string values (but not other sets
+  or dicts).  These must be sorted by integer or byte value for ints or strings, respectively;
+  integers sort before strings if a list contains both.  Sets must contain at least one value: empty
```
> maybe it's better to keep the wording consistent here?

i think so. `set` would be consistent terminology.

fracting commented:
Science fiction: with fully homomorphic encryption, in theory, we can compare sequential numbers on the server side.
Not a serious proposal though, but it would be interesting to come back and re-evaluate the potential of FHE in a couple of years.

ghost commented:
There is an edge case I think the document doesn't explicitly cover:
What happens if a key is removed twice?

Suppose two clients push update with seqno=125 where:
* 125a sets `["int1"]` to `5`, and delete the `"string1"` key at the same time
* 125b sets `["int1"]` to `6`, and delete the `"string1"` key at the same time

By applying simple conflict resolution rule, `["int1"]` will be set to `6` after merging because 125b is higher than 125a.
However, the document doesn't say what will happen to `"string1"` in this case since there are two attempts to remove it.

My understanding is every operation (assign, deletion) should be idempotent, removing the same key twice should be equivalent to removing the same key once. It might seem too obvious so the document doesn't cover it, but it would be nice to have a unit test case on deleting the same key multiple times, so we can make sure the client won't crash when removing a non-exists key.

jagerman committed:
Implement main config message storage structure and diffing

jagerman committed:
Add signature field

Some types of config messages need to be signed (e.g. closed group
config messages) because they will be *placeable* by multiple people
(closed group members), but only actually accepted for a subset
(admins).

jagerman committed:
Add config serialization and signing

jagerman committed:
Config loading, merging, conflict resolution

Splits ConfigMessage into a read-only base class and a
MutableConfigMessage derived class.  You use the latter if you want to
make data changes, the former if you are only reading.

Adds deserialization.

Adds merging and conflict resolution.

Fixes things.

Adds tests implementing the merge logic doc.

jagerman committed:
Add sig verification & optional sig tests

jagerman committed:
Add deterministic config message encryption

jagerman committed:
Use oxenc's _hex for hex literals

jagerman committed:
Working C + C++ implementations of user profiles

jagerman committed:
Update oxen-encoding submodule

jagerman committed:
Bump clang-format version

jagerman committed:
Add lots of comments to example

Also renames the unprefixed `confirm_pushed` C function to
`config_confirm_pushed`.

jagerman committed:
doc wording updates from feedback

jagerman commented:
> There is an edge case I think the document doesn't explicitly cover: What happens if a key is removed twice?

Nothing; removing a key (or set value, for that matter) that doesn't exist does indeed just get ignored.  I've updated the doc to make this explicit.

jagerman committed:
Add ios build script & drone build

jagerman committed:
Various compiler fixes and workarounds

- Add missing header for libc++
- Clang format 15 because 14 is buggy
- Workarounds for compilation failures under macos 10.12, gcc8, and/or
  mingw

jagerman committed:
Fix test suite bug

This isn't a null-terminated C-string.

(How this worked on 64-bit platforms but fails on 32-bit platforms is a
mystery.)

majestrate commented on .drone.jsonnet:
```diff
@@ -252,4 +252,28 @@ local mac_builder(name,
+          'echo "Building on ${DRONE_STAGE_MACHINE}"',
+          // If you don't do this then the C compiler doesn't have an include path containing
+          // basic system headers.  WTF apple:
+          'export SDKROOT="$(xcrun --sdk macosx --show-sdk-path)"',
```
we should also put a precondition in `ios.sh` that makes it die if `$SDKROOT` isnt set.

jagerman replied:
It's up to you to set up your compiler environment properly before compiling things (as is done here for drone). SDKROOT is one such way to do it for stock xcode clang, but if you want to use some other build mechanism then so be it.

jagerman commented:
Merging this now because this is the branch I'm using for other things anyway (i.e. this has effectively become `dev`).
